### PR TITLE
feat(sdk): Expose upload pipeline version

### PR DESCRIPTION
**Pull request by Intellicode**
**Description of your changes:**

Exposes upload pipeline versions 

If possible, can this be merged in the current release branch?

### CODE REVIEW

**Comment by Intellicode** [comment]
@googlebot I signed it

File: sdk/python/kfp/_client.py [hunk 1]
```diff
@@ -741,6 +741,28 @@ def upload_pipeline(
       IPython.display.display(IPython.display.HTML(html))
     return response
 
+  def upload_pipeline_version(
+    self,
+    pipeline_package_path,
+    pipeline_name: str,
```
Comment (Bobgy) [inline sdk/python/kfp/_client.py:747]: nit: should this be pipeline_version_name instead?

File: sdk/python/kfp/_client.py [hunk 2]
```diff
@@ -741,6 +741,28 @@ def upload_pipeline(
...
+    """Uploads a new version of the pipeline to the Kubeflow Pipelines cluster.
+    Args:
+      pipeline_package_path: Local path to the pipeline package.
+      pipeline_name:  Name of the pipeline to be shown in the UI.
+      pipeline_id: Id of the pipeline.
+    Returns:
+      Server response object containing pipleine id and other information.
+    """
+
+    response = self._upload_api.upload_pipeline_version(pipeline_package_path, name=pipeline_name, pipelineid=pipeline_id)
```
Comment (Bobgy) [inline sdk/python/kfp/_client.py:759]: Can we handle errors?
e.g. use a non existent pipeline_id.

**Comment by Bobgy** [comment]
Yes, this is relatively safe to cherry pick.

Thanks for your contribution!

**Comment by Bobgy** [comment]
Hold on, @NikeNano is this feature already supported?
https://github.com/kubeflow/pipelines/pull/4087/files#diff-8866aa3807152db92a11d390f982e484R66

File: sdk/python/kfp/_client.py (hunk 2 above)
Comment (Intellicode) [reply sdk/python/kfp/_client.py:759]: If I look at the other functions in the same file, (for example `upload_pipeline`, they also don't have additional error handling, do you want to add it for this one anyway?

**Comment by Bobgy** [comment]
Can you confirm if your PR was already implemented in https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/cli/pipeline.py#L66?

Can we close?

**Comment by NikeNano** [comment]
> Can you confirm if your PR was already implemented in https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/cli/pipeline.py#L66?
> 
> Can we close?

I think the difference is that this is for the SDK and the merged [PR](https://github.com/kubeflow/pipelines/pull/4087/files#diff-8866aa3807152db92a11d390f982e484R66) is for the CLI.  This functionally now exsist in the CLI put I think it makes sense to also have it in SDK.

**Comment by Intellicode** [comment]
> > Can you confirm if your PR was already implemented in https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/cli/pipeline.py#L66?
> > Can we close?

Indeed, this is for the SDK and not the CLI.

> 
> I think the difference is that this is for the python SDK and the merged [PR](https://github.com/kubeflow/pipelines/pull/4087/files#diff-8866aa3807152db92a11d390f982e484R66) is for the CLI. This functionally now exists in the CLI put I think it makes sense to also have it in SDK.
> 
> UPDATE:This PR is actually mostly similar to the "original" implementation of upload pipeline since it don't allow the name of the pipeline it self to be used, only the pipeline id.

yes that's right, another approach would be to add a parameter to the `upload_pipeline` function to allow a version to be added. However, this maps better to the underying function of the `_upload_api`.

**Comment by NikeNano** [comment]
> > I think the difference is that this is for the python SDK and the merged [PR](https://github.com/kubeflow/pipelines/pull/4087/files#diff-8866aa3807152db92a11d390f982e484R66) is for the CLI. This functionally now exists in the CLI put I think it makes sense to also have it in SDK.
> > UPDATE:This PR is actually mostly similar to the "original" implementation of upload pipeline since it don't allow the name of the pipeline it self to be used, only the pipeline id.
> 
> yes that's right, another approach would be to add a parameter to the `upload_pipeline` function to allow a version to be added. However, this maps better to the underying function of the `_upload_api`.

It might make sens to also allow to pass the pipeline name as an alternative to the pipeline id, you could use the same function as the CLI in that case to get the pipeline id: https://github.com/kubeflow/pipelines/blob/48f3075f0a8cb87e6e5336692f0d9fd4b6150ae9/sdk/python/kfp/_client.py#L315

**Comment by Intellicode** [comment]
> It might make sens to also allow to pass the pipeline name as an alternative to the pipeline id, you could use the same function as the CLI in that case to get the pipeline id:
> 
> https://github.com/kubeflow/pipelines/blob/48f3075f0a8cb87e6e5336692f0d9fd4b6150ae9/sdk/python/kfp/_client.py#L315

Alright, I'll add the parameter as well with pipeline_id given preference when passed

**Comment by Intellicode** [comment]
> Alright, I'll add the parameter as well with pipeline_id given preference when passed

Actually, I've implemented it similar to how the `cron_expression` or `interval_seconds` is required

File: sdk/python/kfp/_client.py [hunk 3]
```diff
@@ -12,27 +12,25 @@
...
 import re
 import tarfile
 import tempfile
+import time
 import warnings
-import yaml
 import zipfile
-import datetime
-from typing import Mapping, Callable
+from typing import Callable, Mapping
```
Comment (NikeNano) [inline sdk/python/kfp/_client.py:25]:
```suggestion
from typing import Callable, Mapping, Optional
```

I think you need to import Optional, which is used on row 777.

Comment (Intellicode) [reply sdk/python/kfp/_client.py]: Woops! I fixed the missing import

**Comment by Bobgy** [comment]
/lgtm
This looks good to me.
/assign @numerology @Ark-kun 
do you want to review?

File: sdk/python/kfp/_client.py [hunk 4]
```diff
@@ -772,6 +772,38 @@ def upload_pipeline(
...
+    """Uploads a new version of the pipeline to the Kubeflow Pipelines cluster.
+    Args:
+      pipeline_package_path: Local path to the pipeline package.
+      pipeline_version_name:  Name of the pipeline version to be shown in the UI.
+      pipeline_id: Optional. Id of the pipeline.
+      pipeline_name: Optional. Name of the pipeline.
+    Returns:
+      Server response object containing pipleine id and other information.
+    Throws:
+      Exception if pipeline id is not found.
```
Comment (numerology) [inline sdk/python/kfp/_client.py:791]: nit: We can be more specific by saying `ValueError when ...`

File: sdk/python/kfp/_client.py [hunk 5]
```diff
@@ -772,6 +772,38 @@ def upload_pipeline(
...
+    """
+
+    if all([pipeline_id, pipeline_name]) or not any([pipeline_id, pipeline_name]):
```
Comment (numerology) [inline sdk/python/kfp/_client.py:794]: nit: We can do `if bool(pipeline_id) == bool(pipeline_name)`, which IMHO is slightly better.

File: sdk/python/kfp/_client.py [hunk 6]
```diff
@@ -772,6 +772,38 @@ def upload_pipeline(
...
+      raise ValueError('Either pipeline_id or pipeline_name is required')
+
+    if pipeline_name:
+      pipeline_id = self.get_pipeline_id(pipeline_name)
+
+    response = self._upload_api.upload_pipeline_version(pipeline_package_path, name=pipeline_version_name, pipelineid=pipeline_id)
```
Comment (numerology) [inline sdk/python/kfp/_client.py:800]: nit: line too long.

File: sdk/python/kfp/_client.py (hunk 5 above)
Comment (Intellicode) [reply sdk/python/kfp/_client.py:795]: I would argue that the original version better communicates the 'all' or 'none' condition of this exception, making it easier to understand. However, if you prefer your approach I'm happy to change it.

Comment (numerology) [reply sdk/python/kfp/_client.py:795]: SG. Up to you

**Pull request merged**